### ISSUE TITLE
Various simulation fixes - [merged]

### ISSUE THREAD

**Issue by SGeeversAtVortech** [Deltares/rtc-tools#1247]
In GitLab by @jvande42b on Apr 26, 2018, 07:46

_Merges various-fixes -> casadi3_

@vreeken I think this fixes the issues I mentioned. A pickled model returns different types than a non-pickled one, so the simulation was initializing incorrectly.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 26, 2018, 07:47

changed target branch from `master` to `casadi3`

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 26, 2018, 07:51

added 11 commits

<ul><li>54270c3e...8e7dbe7c - 7 commits from branch <code>casadi3</code></li><li>04fe1801 - Simulation: rewrite initialization logic to be more tolerant of start attribute types</li><li>38e69974 - Simulation: make initial guess for all states nonzero to avoid invalid number in casadi</li><li>b3380698 - Simulation: implement extract_results() method for simulation, as it is useful…</li><li>fd56b3c9 - Simulation: various comment and whitespace fixes</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/106/diffs?diff_id=13824924&start_sha=54270c3e76d6c7a962dfac914576cc1829ecf19e)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Apr 26, 2018, 11:39

Aside from the tests failing, wouldn't it make more sense to fix this on the pymoca side by making sure the compiled/cached model returns the same types? When you say "compiled" I take it you mean .so/.dll, and not a pickled cache? My guess is that it would affect both, because the variable metadata function is shared between the two.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 27, 2018, 13:53

I meant both pickled and compiled. But I think that the changes here will be more reliable in general.  It is not really pythonic to determine behavior based on type, which is what I was doing before.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Apr 27, 2018, 15:22

Commented on [src/rtctools/simulation/simulation_problem.py line 250](https://github.com/Deltares/rtc-tools/compare/82076448672dadca46a4bf48219d0b3ef6586c95..044f9d2e8f662c0a2fe3814111de8c78186243dd#diff-5ea63c5cd21b27afb74ff666282a3367R250)



Shouldn't this be `start_val is None` now to match the previous behavior? In the new case it seems to me that `0.0` would be a valid value, whereas it was a fallback value previously.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 27, 2018, 15:41

Commented on [src/rtctools/simulation/simulation_problem.py line 250](https://github.com/Deltares/rtc-tools/compare/82076448672dadca46a4bf48219d0b3ef6586c95..044f9d2e8f662c0a2fe3814111de8c78186243dd#diff-5ea63c5cd21b27afb74ff666282a3367R250)



With these changes, I am being more explicit about the distinction between a start value and the `var.start` attribute. If `start_val is None`, that means that var.start contains an expression, which means it is not the default of `0.0`. We only set values from `initial_state()` if the `var.start` and `var.fixed` are default.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 27, 2018, 15:54

Commented on [src/rtctools/simulation/simulation_problem.py line 250](https://github.com/Deltares/rtc-tools/compare/82076448672dadca46a4bf48219d0b3ef6586c95..044f9d2e8f662c0a2fe3814111de8c78186243dd#diff-5ea63c5cd21b27afb74ff666282a3367R250)



The previous behaviour assumed that default values would always be python types, which is not true when loading a cached model. This is why a cached model erroneously behaved differently than an uncached model.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 27, 2018, 17:17

Commented on [src/rtctools/simulation/simulation_problem.py line 336](https://github.com/Deltares/rtc-tools/compare/8e7dbe7c222bd65e68df577064beba2ede585757..fd56b3c926b43ca35f4e8ce105a770b26f806dd8#diff-5ea63c5cd21b27afb74ff666282a3367R336)



@TPiovesan @vreeken I do this because the expression `a = y / x` will fail if the guess for x is `0.0`. But I think that it might be better to pick a number on the order of `1.0e-3`. Alternatively, we can try to set a guess intelligently based on how the value appears in the equations, but I don't know a good way to do that.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 27, 2018, 17:23

Commented on [src/rtctools/simulation/simulation_problem.py line 336](https://github.com/Deltares/rtc-tools/compare/8e7dbe7c222bd65e68df577064beba2ede585757..fd56b3c926b43ca35f4e8ce105a770b26f806dd8#diff-5ea63c5cd21b27afb74ff666282a3367R336)



Are there other cases we should consider when generating a guess? The main thing is that if we pick a common value, e.g. 0.0 or 1.0, there is a good chance the first evaluation of the DAE will hit a undefined point. Another case is `sqrt(-x)` for x in `(-inf, 0.0]`: this would fail currently. Of course, we can oblige the user to either write well-formed models or provide initial conditions. But if there is a nice way we can be a little smarter, this would be a nice improvement.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 27, 2018, 17:37

marked as a **Work In Progress**

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 27, 2018, 17:37

@vreeken This is a WIP because I would like to review the initial guess work a bit more

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Apr 28, 2018, 12:42

Commented on [src/rtctools/simulation/simulation_problem.py line 250](https://github.com/Deltares/rtc-tools/compare/82076448672dadca46a4bf48219d0b3ef6586c95..044f9d2e8f662c0a2fe3814111de8c78186243dd#diff-5ea63c5cd21b27afb74ff666282a3367R250)



I agree with making a distinction between an expression and a default value of 0.0. However, I do want to try to keep/make the cached/compiled models match the freshly parsed models as exactly as possible, so a 4th pass in Pymoca API seems appropriate. Something like this snippet below?

```python
        # Coerce into original type
        for key in variables_with_metadata:
            for i, d in enumerate(db[key]):
                variable = variable_dict[d['name']]
                for j, tmp in enumerate(CASADI_ATTRIBUTES):
                    val = getattr(variable, tmp)
                    if isinstance(val, ca.MX) and val.is_constant():
                        setattr(variable, tmp, variable.python_type(val.to_DM()))
                    elif isinstance(val, ca.DM):
                        setattr(variable, tmp, variable.python_type(val))
```

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 14:49

Commented on [src/rtctools/simulation/simulation_problem.py line 336](https://github.com/Deltares/rtc-tools/compare/8e7dbe7c222bd65e68df577064beba2ede585757..fd56b3c926b43ca35f4e8ce105a770b26f806dd8#diff-5ea63c5cd21b27afb74ff666282a3367R336)



changed this line in [version 4 of the diff](https://gitlab.com/deltares/rtc-tools/merge_requests/106/diffs?diff_id=13954139&start_sha=fd56b3c926b43ca35f4e8ce105a770b26f806dd8#15e34bb5e5838c1e165ddfcd99c9c9596b965eb6_336_334)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 14:49

added 2 commits

<ul><li>ba314989 - Simulation: fix regex so that we eliminate subscripted subcomponents</li><li>9f64153e - Revert &quot;Simulation: make initial guess for all states nonzero to avoid invalid number in casadi&quot;</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/106/diffs?diff_id=13954139&start_sha=fd56b3c926b43ca35f4e8ce105a770b26f806dd8)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 14:49

unmarked as a **Work In Progress**

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 14:51

Commented on [src/rtctools/simulation/simulation_problem.py line 250](https://github.com/Deltares/rtc-tools/compare/82076448672dadca46a4bf48219d0b3ef6586c95..044f9d2e8f662c0a2fe3814111de8c78186243dd#diff-5ea63c5cd21b27afb74ff666282a3367R250)



:+1:  seems reasonable

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 14:52

Commented on [src/rtctools/simulation/simulation_problem.py line 336](https://github.com/Deltares/rtc-tools/compare/8e7dbe7c222bd65e68df577064beba2ede585757..fd56b3c926b43ca35f4e8ce105a770b26f806dd8#diff-5ea63c5cd21b27afb74ff666282a3367R336)



For now, we should just rely on the model either being well-formed or having an initial condition provided.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 15:32

marked as a **Work In Progress**

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 17:08

added 7 commits

<ul><li>9f64153e...b3f098ed - 2 commits from branch <code>casadi3</code></li><li>54f30713 - Simulation: rewrite initialization logic</li><li>2ab185f8 - Simulation: implement extract_results() method for simulation, as it is useful…</li><li>93400f03 - Simulation: various comment and whitespace fixes</li><li>2a79ff00 - Simulation: fix regex so that we also eliminate subcomponents that start with underscore</li><li>9a9924ac - Simulation: use min and max variable attributes in initialize()</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/106/diffs?diff_id=13960134&start_sha=9f64153e8378f440eaf5ab08d1fbd5d8fc1f9f32)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 17:09

unmarked as a **Work In Progress**

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 17:25

added 5 commits

<ul><li>5deb31e3 - Simulation: rewrite initialization logic</li><li>08f1050a - Simulation: implement extract_results() method for simulation, as it is useful…</li><li>88720938 - Simulation: various comment and whitespace fixes</li><li>ac4614c0 - Simulation: fix regex so that we also eliminate subcomponents that start with underscore</li><li>24ffed3a - Simulation: use min and max variable attributes in initialize()</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/106/diffs?diff_id=13960722&start_sha=9a9924acb30c202994c479f778b57f5c146a029a)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 18:48

added 5 commits

<ul><li>651098ca - Simulation: rewrite initialization logic</li><li>5ebec339 - Simulation: implement extract_results()</li><li>de268ac4 - Simulation: various comment and whitespace fixes</li><li>4dc2838c - Simulation: also eliminate subcomponents that start with underscore</li><li>ae016184 - Simulation: use min and max variable attributes in initialize()</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/106/diffs?diff_id=13963411&start_sha=24ffed3a7bb66f88d9bc31d689f46f50331661a8)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on Apr 30, 2018, 18:48

resolved all discussions

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Apr 30, 2018, 19:23

added 6 commits

<ul><li>82076448 - 1 commit from branch <code>casadi3</code></li><li>607a8a3f - Simulation: rewrite initialization logic</li><li>ba32ffc0 - Simulation: implement extract_results()</li><li>e6745c2b - Simulation: various comment and whitespace fixes</li><li>a3469e51 - Simulation: also eliminate subcomponents that start with underscore</li><li>044f9d2e - Simulation: use min and max variable attributes in initialize()</li></ul>

[Compare with previous version](https://gitlab.com/deltares/rtc-tools/merge_requests/106/diffs?diff_id=13964304&start_sha=ae016184de343b43e573f22277a18863ddb39a8f)

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Apr 30, 2018, 19:36

merged
